src/makeguids.h

```diff
@@ -0,0 +1,277 @@
+	 *    guids in place
+	 *
+	 * 4) move the first entry to the end of guids
+	 *
+	 * 5) store a struct strtab in the first position
+	 *
+	 * 6) realloc guids and strtab down to size
+	 *
+	 * 8) sort guids[1:n]
+	 *
+	 * 9) set guidnames = guids
+	 *
+	 */
+
+	/*
+	 * pad (at least) one NUL char at the end.
+	 */
+	char *nbuf = realloc(buf, ALIGN_UP((bufsz+1), page_size));
+	if (!nbuf) {
+		efi_error("Could not realloc guid file buffer");
+		goto err_free_buf;
+	}
+	buf = nbuf;
+	buf[bufsz] = '\0';
+
+	struct guidname_offset *guids = NULL;
+	guids = calloc(1, ALIGN_UP(bufsz, page_size));
+	if (!guids)
+		goto err_free_buf;
+
+	char *strtab = calloc(1, ALIGN_UP(bufsz, page_size));
+	if (!strtab)
+		goto err_free_guids;
+
+	size_t bpos = 0, spos = 0, nguids = 0;
+	for (; bpos < bufsz && buf[bpos] != 0; nguids++) {
+		struct guidname_offset *gno = &guids[nguids];
+		char guidstr[37];
+
+		char *symbol = strchr(&buf[bpos], '\t');
+		if (symbol == NULL) {
+			efi_error("invalid guid string data on line %zd", nguids);
+			goto err_free_strtab;
+		}
+		*symbol = '\0';
+		symbol += 1;
+
+		strncpy(guidstr, &buf[bpos], 37);
+		guidstr[36] = '\0';
+
+		char *desc = strchrnul(symbol, '\t');
+		if (*desc != '\0') {
+			*desc = '\0';
+			desc += 1;
+		}
+
+		char *end = strchrnul(desc, '\n');
+		*end = '\0';
+
+		efi_guid_t guid;
+		rc = efi_str_to_guid_(guidstr, &guid);
+		if (rc < 0) {
+			efi_error("unparsable guid on line %zd", nguids);
+			goto err_free_strtab;
+		}
+
+		char *s = &strtab[spos];
+		off_t nameoff, symoff, descoff;
+
+		nameoff = s - strtab;
+		s = stpcpy(s, symbol);
+		*(s++) = '\0';
+		symoff = s - strtab;
+		s = stpcpy(s, "efi_guid_");
+		s = stpcpy(s, symbol);
+		*(s++) = '\0';
+		descoff = s - strtab;
+		s = stpcpy(s, desc);
+		*(s++) = '\0';
+		spos = s - strtab;
+
+		gno->guid = guid;
+		gno->nameoff = nameoff;
+		gno->symoff = symoff;
+		gno->descoff = descoff;
+		bpos = end - buf + 1;
+	}
+
+	if (spos == 0) {
+		efi_error("guid table produced no strings");
+		goto err_free_strtab;
+	}
+
+	char *newstrtab = realloc(strtab, ALIGN_UP(spos, page_size));
+	if (!newstrtab) {
+		efi_error("could not realloc strtab for guid table");
+		goto err_free_strtab;
+	}
+	strtab = newstrtab;
+
+	struct guidname_offset *newguids;
+	newguids = realloc(guids,
+			   ALIGN_UP((1+nguids) * sizeof(newguids[0]),
+				    page_size));
+	if (!newguids) {
+		efi_error("could not realloc guidnames table");
+		goto err_free_strtab;
+	}
+	guids = newguids;
+
+	memcpy(&guids[nguids], &guids[0], sizeof(guids[0]));
+
+	struct guidname_index *guidnames;
+	guidnames = (struct guidname_index *)guids;
+
+	guidnames->strtab = strtab;
+	guidnames->nguids = nguids;
+	guidnames->strsz = spos;
+
+	qsort_r(&guidnames->offsets[0], nguids,
+		sizeof(struct guidname_offset), gnopguidcmp, guidnames);
+
+	*guidnamesp = guidnames;
+
+	xfree(buf);
+	return 0;
+
+err_free_strtab:
+	xfree(strtab);
+err_free_guids:
+	xfree(guids);
+err_free_buf:
```

Comment:
If you NULL-initialize these at the top, you can have one labeled exit instead of three.